pkg/netflow/goflowlib/flowroutine.go

```diff
@@ -0,0 +1,96 @@
+package goflowlib
+
+import (
+	"github.com/netsampler/goflow2/format"
+	"github.com/netsampler/goflow2/utils"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+// setting reusePort to false since not expected to be useful
+// more info here: https://stackoverflow.com/questions/14388706/how-do-so-reuseaddr-and-so-reuseport-differ
+const reusePort = false
+
+// FlowRoutineState is a wrapper for StateNetFlow/StateSFlow/StateNFLegacy
+type FlowRoutineState struct {
+	state    interface{}
+	hostname string
+	port     uint16
+}
+
+// Shutdown is a wrapper for StateNetFlow/StateSFlow/StateNFLegacy Shutdown method
+func (s *FlowRoutineState) Shutdown() {
+	switch state := s.state.(type) {
+	case *utils.StateNetFlow:
+		log.Infof("Shutdown NetFlow9/ipfix listener on %s:%d", s.hostname, s.port)
+		state.Shutdown()
+	case *utils.StateSFlow:
+		log.Infof("Shutdown sFlow listener on %s:%d", s.hostname, s.port)
+		state.Shutdown()
+	case *utils.StateNFLegacy:
+		log.Infof("Shutdown Netflow5 listener on %s:%d", s.hostname, s.port)
+		state.Shutdown()
+	default:
+		log.Warnf("Unknown flow listener state type `%T` for %s:%d", s.hostname, s.port)
+	}
+}
+
+// StartNetFlowRoutine returns a FlowRoutineState for StateNetFlow
```

Comment:
<!-- Thoughts represent an idea that popped up from reviewing. These comments are non-blocking by nature, but they are extremely valuable and can lead to more focused initiatives and mentoring opportunities.  -->
**💭 thought:** ‏The next three functions looks quite duplicated... Maybe we can factorize?

Reply:
The issue is that `utils.StateNFLegacy/StateNetFlow/StateSFlow` don't currently share a common interface that expose `FlowRoutine` method.

I think we can improve that by possibly contributing upstream to improve the API e.g. by adding a common interface.

Right now, IMHO the duplication is still reasonable and very localised.